File: ios/Sasquatsh/ViewModels/RaffleViewModel.swift
import SwiftUI

@Observable
@MainActor
final class RaffleViewModel {
    var raffle: Raffle?
    var isLoading = false
    var error: String?
    var mailInName = ""
    var mailInAddress = ""
    var isSubmittingMailIn = false
    var mailInSuccess = false

    private var services: ServiceContainer?

    func configure(services: ServiceContainer) {
        self.services = services
    }

    func loadActiveRaffle() async {
        guard let services else { return }
        isLoading = true
        error = nil
        raffle = try? await services.raffle.getActiveRaffle()
        isLoading = false
    }

    func submitMailInEntry() async {
        guard let services, let raffle else { return }
        let name = mailInName.trimmingCharacters(in: .whitespacesAndNewlines)
        let address = mailInAddress.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !name.isEmpty, !address.isEmpty else {
            error = "Name and address are required"
            return
        }
        isSubmittingMailIn = true
        error = nil
        do {
            try await services.raffle.submitMailInEntry(raffleId: raffle.id, name: name, address: address)
            mailInSuccess = true
            mailInName = ""
            mailInAddress = ""
            await loadActiveRaffle()
        } catch {
            self.error = error.localizedDescription
        }
        isSubmittingMailIn = false
    }

    var entriesByType: [(type: RaffleEntryType, count: Int)] {
        guard let entries = raffle?.userEntries else { return [] }
        var result: [(RaffleEntryType, Int)] = []
        for entryType in RaffleEntryType.allCases {
            let count = entries.filter { $0.entryType == entryType.rawValue }
                .reduce(0) { $0 + $1.entryCount }
            if count > 0 {
                result.append((entryType, count))
            }
        }
        return result
    }
}

File: android/app/src/main/java/com/sasquatsh/app/viewmodels/RaffleViewModel.kt
package com.sasquatsh.app.viewmodels

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.sasquatsh.app.models.Raffle
import com.sasquatsh.app.models.RaffleEntryType
import com.sasquatsh.app.services.RaffleService
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

data class RaffleUiState(
    val raffle: Raffle? = null,
    val isLoading: Boolean = false,
    val error: String? = null,
    val mailInName: String = "",
    val mailInAddress: String = "",
    val isSubmittingMailIn: Boolean = false,
    val mailInSuccess: Boolean = false
) {
    val entriesByType: List<Pair<RaffleEntryType, Int>>
        get() {
            val entries = raffle?.userEntries ?: return emptyList()
            return RaffleEntryType.values().mapNotNull { entryType ->
                val count = entries
                    .filter { it.entryType == entryType.value }
                    .sumOf { it.entryCount }
                if (count > 0) entryType to count else null
            }
        }
}

@HiltViewModel
class RaffleViewModel @Inject constructor(
    private val raffleService: RaffleService
) : ViewModel() {

    private val _uiState = MutableStateFlow(RaffleUiState())
    val uiState: StateFlow<RaffleUiState> = _uiState.asStateFlow()

    fun loadActiveRaffle() {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true, error = null) }
            val raffle = try {
                raffleService.getActiveRaffle()
            } catch (_: Exception) {
                null
            }
            _uiState.update { it.copy(raffle = raffle, isLoading = false) }
        }
    }

    fun updateMailInName(name: String) {
        _uiState.update { it.copy(mailInName = name) }
    }

    fun updateMailInAddress(address: String) {
        _uiState.update { it.copy(mailInAddress = address) }
    }

    fun submitMailInEntry() {
        val raffle = _uiState.value.raffle ?: return
        val name = _uiState.value.mailInName.trim()
        val address = _uiState.value.mailInAddress.trim()

        if (name.isEmpty() || address.isEmpty()) {
            _uiState.update { it.copy(error = "Name and address are required") }
            return
        }

        viewModelScope.launch {
            _uiState.update { it.copy(isSubmittingMailIn = true, error = null) }
            try {
                raffleService.submitMailInEntry(raffle.id, name, address)
                _uiState.update {
                    it.copy(
                        mailInSuccess = true,
                        mailInName = "",
                        mailInAddress = "",
                        isSubmittingMailIn = false
                    )
                }
                loadActiveRaffle()
            } catch (e: Exception) {
                _uiState.update {
                    it.copy(error = e.localizedMessage, isSubmittingMailIn = false)
                }
            }
        }
    }

    fun clearError() {
        _uiState.update { it.copy(error = null) }
    }
}
